File: discord-bot-py/src/nextbot/encrypt.py
import secrets


MYBASE16_CHARSET = "0123456789, "
BASE64_CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"


def _require_characters(value: str, charset: str, name: str) -> None:
    invalid_characters = set(value) - set(charset)
    if invalid_characters:
        raise ValueError(f"{name} contains unsupported characters")


def _base16_to_64(value: str) -> str:
    output: list[str] = []
    for index in range(0, len(value), 3):
        target = value[index:index + 3].ljust(3)
        number = (
            MYBASE16_CHARSET.index(target[0]) * 16 * 16
            + MYBASE16_CHARSET.index(target[1]) * 16
            + MYBASE16_CHARSET.index(target[2])
        )
        output.append(BASE64_CHARSET[(number >> 6) & 0x3F])
        output.append(BASE64_CHARSET[number & 0x3F])
    return "".join(output)


def _base64_to_16(value: str) -> str:
    output: list[str] = []
    for index in range(0, len(value), 2):
        target = value[index:index + 2].ljust(2, "A")
        number = BASE64_CHARSET.index(target[0]) * 64 + BASE64_CHARSET.index(target[1])
        decoded_indexes = (
            (number >> 8) & 0xFF,
            (number >> 4) & 0x0F,
            number & 0x0F,
        )
        if any(decoded_index >= len(MYBASE16_CHARSET) for decoded_index in decoded_indexes):
            raise ValueError("ciphertext was not encrypted with the supplied key")
        output.extend(MYBASE16_CHARSET[decoded_index] for decoded_index in decoded_indexes)
    return "".join(output)


def encrypt(plaintext: str, key: str) -> str:
    _require_characters(plaintext, MYBASE16_CHARSET, "plaintext")
    _require_characters(key, BASE64_CHARSET, "key")
    if plaintext and not key:
        raise ValueError("key must not be empty")

    random_prefix = generate_key(1)

    encoded = _base16_to_64(plaintext)
    output: list[str] = []
    previous_index = BASE64_CHARSET.index(random_prefix)
    output.append(random_prefix)
    for index, character in enumerate(encoded):
        character_index = BASE64_CHARSET.index(character)
        key_index = BASE64_CHARSET.index(key[index % len(key)])
        output.append(BASE64_CHARSET[(character_index + key_index + previous_index) % 64])
        previous_index = character_index
    return "".join(output)


def decrypt(ciphertext: str, key: str) -> str:
    _require_characters(ciphertext, BASE64_CHARSET, "ciphertext")
    _require_characters(key, BASE64_CHARSET, "key")
    if ciphertext and not key:
        raise ValueError("key must not be empty")

    decoded: list[str] = []
    previous_index = BASE64_CHARSET.index(ciphertext[0])
    ciphertext = ciphertext[1:]  # Remove the random prefix
    for index, character in enumerate(ciphertext):
        character_index = BASE64_CHARSET.index(character)
        key_index = BASE64_CHARSET.index(key[index % len(key)])
        decoded_index = (character_index - key_index - previous_index) % 64
        decoded.append(BASE64_CHARSET[decoded_index])
        previous_index = decoded_index
    return _base64_to_16("".join(decoded))


def generate_key(length: int) -> str:
    if length < 0:
        raise ValueError("length must not be negative")
    return "".join(secrets.choice(BASE64_CHARSET) for _ in range(length))

File: web-app-ts/src/utils/encrypt.ts

const MYBASE16_CHARSET = '0123456789, ';
const BASE64_CHARSET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

function base16to64(input: string): string {
  let output = '';
  for (let i = 0; i < input.length; i += 3) {
    let number = 0;
    const target = input.substring(i, i + 3) + '  '.substring(0, 3 - (input.length - i));

    number += MYBASE16_CHARSET.indexOf(target[0]) * 16 * 16;
    number += MYBASE16_CHARSET.indexOf(target[1]) * 16;
    number += MYBASE16_CHARSET.indexOf(target[2]);

    output += BASE64_CHARSET[(number >> 6) & 0x3f];
    output += BASE64_CHARSET[number & 0x3f];
  }
  
  return output;
}

function base64to16(input: string): string {
  let output = '';
  for (let i = 0; i < input.length; i += 2) {
    let number = 0;
    const target = input.substring(i, i + 2) + 'A'.substring(0, 2 - (input.length - i));

    number += BASE64_CHARSET.indexOf(target[0]) * 64;
    number += BASE64_CHARSET.indexOf(target[1]);

    output += MYBASE16_CHARSET[(number >> 8) & 0xff];
    output += MYBASE16_CHARSET[(number >> 4) & 0xf];
    output += MYBASE16_CHARSET[number & 0xf];
  }

  return output;
}

//#endregion
export function encrypt(plaintext : string, key: string): string {
  const keyBase64 = base16to64(plaintext);
  let output = '';
  let previousIndex = 0;
  for (let i = 0; i < keyBase64.length; i++) {
    const keyChar = key[i % key.length];
    const keyIndex = BASE64_CHARSET.indexOf(keyChar);
    const characterIndex = BASE64_CHARSET.indexOf(keyBase64[i]);
    output += BASE64_CHARSET[(characterIndex + keyIndex + previousIndex) % 64];
    previousIndex = characterIndex;
  }

  return output;
}

export function decrypt(ciphertext: string, key: string): string {
  let output = '';
  let previousIndex = 0;
  for (let i = 0; i < ciphertext.length; i++) {
    const keyChar = key[i % key.length];
    const keyIndex = BASE64_CHARSET.indexOf(keyChar);
    const characterIndex = (BASE64_CHARSET.indexOf(ciphertext[i]) - keyIndex - previousIndex + 64 * 2) % 64;
    output += BASE64_CHARSET[characterIndex];
    previousIndex = characterIndex;
  }

  return base64to16(output);
}

export function generateKey(length: number): string {
  let key = '';
  for (let i = 0; i < length; i++) {
    const randomIndex = Math.floor(Math.random() * BASE64_CHARSET.length);
    key += BASE64_CHARSET[randomIndex];
  }
  return key;
}
